### ISSUE TITLE
Incorrect errors for Array.indexOf/push/unshift/splice in the case of generic method unions 

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#57638]
### 🔎 Search Terms

Argument of type 'string | number' is not assignable to parameter of type 'never'. Type 'string' is not assignable to type 'never'.ts(2345)

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/FAEwpgxgNghgTmABAN3ogngLkQQTnGdAHgGcAXOASwDsBzAPgB88DjqBXAWwCMw56A3MFCRYCRBAD21cohjYOPPo3JU6Q4OgB0NcAA8A8gDMAFDACUAxAHpriGlPyQyiPnElxh2gA7sSACzNLGzsHDwQIFzcPLy12GX9KIzIgq1t7akcIqPwYzS0SbyhKCDATAEYAGiqLNNDM8OdXXM9hYSA

### 💻 Code

```ts
declare var y: Array<string>|Array<number>;

declare const a: number|string;

y.indexOf(a); // incorrect error on a 

y.push(a); // incorrect error on a

y.unshift(a); // incorrect error on a

y.splice(1,1,a); // incorrect error on a

```

### 🙁 Actual behavior

Argument of type 'string | number' is not assignable to parameter of type 'never'. Type 'string' is not assignable to type 'never'.ts(2345)

### 🙂 Expected behavior

No error

### Additional information about the issue

Fix available pull #57636

**Comment by RyanCavanaugh** [comment]
Issuing an error if you're `push`ing a value that might be a `number` into an array that might be a `string[]` is the correct behavior.
